retry/retry.go

```diff
@@ -0,0 +1,262 @@
+	// Next will return immediately and subsequent calls will behave as if
+	// they had followed the very first attempt.
+	Reset()
+
+	// Next returns whether the retry loop should continue, and blocks for the
+	// appropriate length of time before yielding back to the caller.
+	//
+	// If a context is present, Next will eagerly return false if the context is done.
+	Next() bool
+
+	// CurrentAttempt returns current retry attempt.
+	//
+	// First attempt number is 0.
+	//
+	// Resetting retrier resets attempts counter.
+	CurrentAttempt() int
+}
+
+// Option configures retry strategy such as backoff duration or maximal number of attempts.
+type Option func(r *options)
+
+// WithMaxAttempts sets upper limit on a number of attempts.
+//
+// Max attempts of 0 indicates no limit.
+//
+// If max attempts option is not used, then default value of 0 is used.
+func WithMaxAttempts(maxAttempts int) Option {
+	return func(o *options) {
+		o.maxAttempts = maxAttempts
+	}
+}
+
+// WithInitialBackoff sets initial backoff.
+//
+// If initial backoff option is not used, then default value of 50 milliseconds is used.
+func WithInitialBackoff(initialBackoff time.Duration) Option {
+	return func(o *options) {
+		o.initialBackoff = initialBackoff
+	}
+}
+
+// WithMaxBackoff sets upper limit on backoff duration.
+//
+// Max backoff of 0 indicates no limit.
+//
+// If max backoff option is not used, then default value of 2 seconds is used.
+func WithMaxBackoff(maxBackoff time.Duration) Option {
+	return func(o *options) {
+		o.maxBackoff = maxBackoff
+	}
+}
+
+// WithMultiplier sets backoff multiplier controlling how fast
+// backoff duration grows with each retry attempt.
+//
+// If multiplier option is not used, then default value of 2 is used.
+func WithMultiplier(multiplier float64) Option {
+	return func(o *options) {
+		o.multiplier = multiplier
+	}
+}
+
+// WithRandomizationFactor sets randomization factor.
+//
+// If randomization factor option is not used, then default value of 0.15 is used.
+func WithRandomizationFactor(randomizationFactor float64) Option {
+	return func(o *options) {
+		o.randomizationFactor = randomizationFactor
+	}
+}
+
+// Start returns a new initialized retrier.
+//
+// If the provided context is canceled (see Context.Done), then Next() will eagerly return false and
+// the retry loop will do no iterations.
+func Start(ctx context.Context, opts ...Option) Retrier {
+	r := &retrier{
+		options: options{
+			maxAttempts:         defaultMaxAttempts,
+			initialBackoff:      defaultInitialBackoff,
+			maxBackoff:          defaultMaxBackoff,
+			multiplier:          defaultMultiplier,
+			randomizationFactor: defaultRandomizationFactor,
+		},
+		ctxDoneChan:    ctx.Done(),
+		currentAttempt: 0,
+		isReset:        false,
+	}
+	for _, option := range opts {
+		option(&r.options)
+	}
+	r.Reset()
+	return r
+}
+
+const (
+	defaultMaxAttempts         = 0 // Infinite retries.
```

Comment:
I do not like this default value (it blocks forever), but worried that changing it would be a break when we have internal consumers move to this import.